### PR TITLE
Allow generic reference type for subimage

### PR DESCRIPTION
**Pull request by HeroicKatora**
As a proposed solution to #762 , this allows any reference type as the underlying storage to `SubImage`. This makes it possible to use the non-mutable `&ImageBuffer`. Only mutable reference types implement the `GenericImage` trait.

### CODE REVIEW

File: src/image.rs [hunk 1]
```diff
@@ -495,33 +496,40 @@ pub trait GenericImage: Sized {
...
-        <Self::Pixel as Pixel>::Subpixel: 'static,
-        Self::Pixel: 'static,
-    {
+    fn sub_image(&mut self, x: u32, y: u32, width: u32, height: u32)
+    -> SubImage<&mut Self>
+    where 
+		Self: 'static, 
+		<Self::Pixel as Pixel>::Subpixel: 'static,
+    	Self::Pixel: 'static 
+	{
```
Comment (bvssvni) [inline src/image.rs]: Can you move this `{` to the beginning of the line?

File: src/image.rs [hunk 2]
```diff
@@ -495,33 +496,40 @@ pub trait GenericImage: Sized {
...
+        SubImage::new(self, x, y, width, height)
+    }
+
+    /// Returns an subimage that is an immutable view into this image.
+    fn view(&self, x: u32, y: u32, width: u32, height: u32)
+    -> SubImage<&Self>
```
Comment (bvssvni) [inline src/image.rs]: This could be put after `)` on previous line.

File: src/image.rs [hunk 3]
```diff
@@ -495,33 +496,40 @@ pub trait GenericImage: Sized {
...
+    	Self::Pixel: 'static 
+	{
         SubImage::new(self, x, y, width, height)
     }
 }
 
 /// A View into another image
-pub struct SubImage<'a, I: 'a> {
-    image: &'a mut I,
+pub struct SubImage <I> {
```
Comment (bvssvni) [inline src/image.rs]: `SubImage<I>`

File: src/image.rs [hunk 4]
```diff
@@ -495,33 +496,40 @@ pub trait GenericImage: Sized {
...
+    image:   I,
```
Comment (bvssvni) [inline src/image.rs]: `image: I,`

File: src/image.rs (hunk 1 above)
Comment (HeroicKatora) [reply src/image.rs]: I just copied that style from `copy_from` directly above. I also think it is slightly more readable due to the visible space between the type bounds and the function implementation.

**Comment by d-dorazio** [comment]
I like this implementation a lot, but I think we could do better here. The problem I see is that I'm basically unable to call any non mutating `GenericImage` functions on `SubImage` because we can only provide an implementation for `&mut I` since `GenericImage` also requires `get_pixel_mut` and `put_pixel`. This is a bit inconvenient imho because we'd be forcing a call to `to_image` to copy the whole view just to call some non mutating methods.

My proposal would be to create a new `GenericImageView` trait that contains only the non mutating methods and then `GenericImage` would also require `GenericImageView`. This way we could implement `GenericImageView` safely for `SubImage<&Self>` and in addition we could still implement `GenericImage` for `SubImage<&mut Self>`. I think that this shouldn't break any code, but I'm not totally sure.

We could also implement this in another PR if you think it's a bit out of scope in this PR since it requires changing `GenericImage`.

@HeroicKatora @bvssvni what do you think?

**Comment by bvssvni** [comment]
@d-dorazio Sounds like a good idea. If this is a weakness in the current API then a breaking change is reasonable.

We won't publish yet, so this could be implemented in another PR.

**Comment by bvssvni** [comment]
Merging.

**Comment by bvssvni** [comment]
Thanks!

**Pull request merged**